Sources/Subjects/ReplaySubject.swift

```diff
@@ -0,0 +1,171 @@
+/// future subscribers and also forwards completion events.
+///
+/// The implementation borrows heavily from [Tristan’s](https://github.com/tcldr/Entwine/blob/b839c9fcc7466878d6a823677ce608da998b95b9/Sources/Entwine/Operators/ReplaySubject.swift)
+/// with the main modifications being leaning on Shai’s [DemandBuffer](https://github.com/CombineCommunity/CombineExt/blob/7ed4677e33fdc963af404423cb563e133c271d2b/Sources/Common/DemandBuffer.swift)
+/// instead of `Entwine`’s [SinkQueue](https://github.com/tcldr/Entwine/blob/b839c9fcc7466878d6a823677ce608da998b95b9/Sources/Common/Utilities/SinkQueue.swift)
+/// and a plain ol’ `[Output]` buffer instead of a
+/// [custom, LinkedListQueue-backed one](https://github.com/tcldr/Entwine/blob/8be24a59bc91410bb29e84b1c4ae35398a5839c8/Sources/Entwine/Operators/ReplaySubject.swift#L162-L177).
+public final class ReplaySubject<Output, Failure: Error> {
+    // MARK: - Private
+
+    private let maxBufferSize: UInt
+    private var buffer = [Output]()
+
+    // MARK: - Internal
+
+    var subscriptions = [Subscription<AnySubscriber<Output, Failure>>]()
+
+    // We also track subscriber identifiers, to more quickly bottom-out double subscribes instead of having to do a
+    // linear pass over `subscriptions`.
+    var subscriberIdentifiers = Set<CombineIdentifier>()
+
+    private var completion: Subscribers.Completion<Failure>?
+    private var isActive: Bool { completion == nil }
+
+    /// The initialization point for `ReplaySubject`s.
+    /// - Parameter maxBufferSize: The maximum number of value events to buffer and replay to all future subscribers.
+    public init(maxBufferSize: UInt) {
+        self.maxBufferSize = maxBufferSize
+    }
+}
+
+// MARK: - `Subject`
+
+extension ReplaySubject: Subject {
+    public func send(_ value: Output) {
+        guard isActive else { return }
+
+        buffer.append(value)
+
+        if buffer.count > maxBufferSize {
+            buffer.removeFirst()
+        }
+
+        subscriptions.forEach { $0.forwardValueToBuffer(value) }
+    }
+
+    public func send(completion: Subscribers.Completion<Failure>) {
+        guard isActive else { return }
+
+        self.completion = completion
+
+        subscriptions.forEach { $0.forwardCompletionToBuffer(completion) }
+    }
+
+    public func send(subscription: Combine.Subscription) {
+        subscription.request(.unlimited)
+    }
+}
+
+// MARK: - `Publisher`
+
+extension ReplaySubject: Publisher {
+    public typealias Output = Output
+    public typealias Failure = Failure
+
+    public func receive<Subscriber: Combine.Subscriber>(
+        subscriber: Subscriber
+    ) where Failure == Subscriber.Failure, Output == Subscriber.Input {
+        let subscriberIdentifier = subscriber.combineIdentifier
+
+        guard !subscriberIdentifiers.contains(subscriberIdentifier) else {
+            subscriber.receive(subscription: Subscriptions.empty)
+            subscriber.receive(completion: .finished)
```

Comment:
is it correct sending a finishing event here?

Reply:
mm yea, actually wondering about this too now (cc @tcldr)—when double subscribing on a `Passthrough-` or `CurrentValueSubject`, I’m not seeing any added terminal events. So, I’m guessing passing down an empty subscription might be all that we should do, too.

Reply:
Yes, just confirmed this myself – thanks for the heads-up! Certainly best to try and maintain consistency with the behaviour that Apple has implemented in their own Subjects.